frame/pablo/src/lib.rs

```diff
@@ -745,22 +697,33 @@ pub mod pallet {
 			}
 		}
 
-		fn amount_of_lp_token_for_added_liquidity(
+		fn add_liquidity_dryrun(
+			who: &Self::AccountId,
 			pool_id: Self::PoolId,
 			base_amount: Self::Balance,
 			quote_amount: Self::Balance,
 		) -> Result<Self::Balance, DispatchError> {
 			let pool = Self::get_pool(pool_id)?;
 			let pool_account = Self::account_id(&pool_id);
 			match pool {
-				PoolConfiguration::StableSwap(pool) =>
-					StableSwap::<T>::calculate_mint_amount_and_fees(
-						&pool,
-						&pool_account,
-						&base_amount,
-						&quote_amount,
-					)
-					.map(|x| x.0),
+				PoolConfiguration::StableSwap(pool) => {
+					let (amount_of_lp_token_to_mint, ..) =
+						StableSwap::<T>::calculate_mint_amount_and_fees(
+							&pool,
+							&pool_account,
+							&base_amount,
+							&quote_amount,
+						)?;
+					ensure!(
+						T::Assets::reducible_balance(pool.pair.base, who, false) >= base_amount,
+						Error::<T>::NotEnoughLiquidity
+					);
+					ensure!(
+						T::Assets::reducible_balance(pool.pair.quote, who, false) >= quote_amount,
+						Error::<T>::NotEnoughLiquidity
+					);
+					Ok(amount_of_lp_token_to_mint)
+				},
```

Comment:
I'd prefer for this to be methods on pool, so that the body of this function is reduced:

```
impl PoolConfiguration {
    // Returns number of LP tokens for the given pool
    fn lp_for_liquidity(...) -> Balance
}
```

Then the body can be:

```
let pool = Self::get_pool(pool_id)?;
let pool_account = Self::account_id(&pool_id);
ensure!(T::Assets::reducible_balance(pool.pair.base, who, false) >= base_amount,
Error::<T>::NotEnoughLiquidity
);
ensure!(
T::Assets::reducible_balance(pool.pair.quote, who, false) >= quote_amount,
Error::<T>::NotEnoughLiquidity
);
let lp = pool.lp_for_liquidity(base_amount, quote_amount);
```